### PR TITLE
protect against corruption

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 when serializing state into an account whose data is too small, the account is corrupted and thereafter unrecoverable

 #### Summary of Changes
 don't write unless it looks like it'll fit

Fixes #

### CODE REVIEW

File: sdk/src/account.rs [hunk 1]
```diff
@@ -65,6 +65,9 @@ impl Account {
     }
 
     pub fn serialize_data<T: serde::Serialize>(&mut self, state: &T) -> Result<(), bincode::Error> {
+        if bincode::serialized_size(state)? > self.data.len() as u64 {
+            return Err(Box::new(bincode::ErrorKind::SizeLimit));
+        }
```
Comment (sagar-solana) [inline sdk/src/account.rs]: So I was wondering about this. According to bincode's serialize, it shouldn't be an issue. Why did we conclude it was a problem again?

https://docs.rs/bincode/0.8.0/bincode/fn.serialize_into.html

Comment (rob-solana) [reply sdk/src/account.rs]: I thought you'd had a problem.  I'd had a problem: when I supplied 1 fewer byte than required for a vector of (u64, Hash).

Comment (rob-solana) [reply sdk/src/account.rs]: I'm looking at bincode now, and it looks like the check is in place if config has a limit in place, but the default bincode config is unlimited, so the check is skipped?

https://github.com/servo/bincode/blob/e84edd4b0ec2b4e5ebf70ea1508cce8db950fdcd/src/internal.rs#L20

https://github.com/servo/bincode/blob/e84edd4b0ec2b4e5ebf70ea1508cce8db950fdcd/src/config.rs#L98

Comment (sagar-solana) [reply sdk/src/account.rs]: Yeah I remember having a problem with this but decided to just take another look at it and saw that documentation. But you're right, it seems we're responsible for setting the limit there.

Comment (sagar-solana) [reply sdk/src/account.rs]: so two ways to do it, your way or pass in the account.data.len() as the size limit?

Comment (rob-solana) [reply sdk/src/account.rs]: bincode::config().limit(account.data.len() as u64).serialize_into() is uglier, IMHO

**Pull request merged**